### PR TITLE
Document our secrets and how to rotate them

### PR DESCRIPTION
**Pull request by barrucadu**
[Trello card](https://trello.com/c/I78BCcih/580-document-our-use-of-secrets)

### CODE REVIEW

File: source/support/rotate-secrets/index.html.md.erb [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+#### `oidc-pepper-production` and `oidc-pepper-staging`
+
+Generate an arbitrary string with `rake secret`.
+
+This secret is used to [generate subject identifiers][], and so
+rotating it will cause services to see a new subject identifier for a
+returning user.  This means that services won't be able to identify
+returning users, and will treat them as new users.
+
+If this secret needs to be rotated, we will need to engage with other
```
Comment (huwd) [inline source/support/rotate-secrets/index.html.md.erb:139]: This is a really good point and one we should cover in our "team docs" vs "gov.uk" docs chat tomorrow.
If we were going to have a "how to rotate the Brexit checker secrets" doc that would sit well in the GOV.UK developer docs.

I guess when we integrate more services we'll need some kind of process for handing this with lots of different support functions 🤔 or we'd break a ton of apps overnight...

File: source/support/rotate-secrets/index.html.md.erb [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+Generate an arbitrary string with `rake secret`.
+
+This secret is used to sign session cookies, and so rotating it will
+log out all users.
+
+#### `zendesk-api-username` and `zendesk-api-key`
+
+Inform User Support that our Zendesk credentials have been leaked, and
+that we need new ones.
+
```
Comment (huwd) [inline source/support/rotate-secrets/index.html.md.erb:180]:
```suggestion

#### `grafana-api-key`

This secret is used to post annotations to our grafana boards to track deploys.
Inform #﻿re-autom8, Accounts team tech lead, or another person with PaaS grafana admin access to recind the API token and generate a new one
```

Comment (huwd) [reply source/support/rotate-secrets/index.html.md.erb:180]: I've just added another one 😅

File: source/support/rotate-secrets/index.html.md.erb [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+# Rotate a secret
+
+If a secret expires or is compromised, you will need to rotate it.
+This is a currently a manual process.
+
+This content explains how secrets are used, what secrets we have, and
+how to rotate them should the need arise.
+
+If a secret is compromised, this is a security incident.  Follow [the
```
Comment (huwd) [inline source/support/rotate-secrets/index.html.md.erb]: I think this is correct and we should chat more in slack about the incident process.
But is there any sense that some secrets are more urgent / bigger security issues than others?

Or is it an incident for everything and we sort out the P score in the review?

Comment (barrucadu) [reply source/support/rotate-secrets/index.html.md.erb]: Some secrets are worse than others, for example if our PaaS credentials are leaked someone could use those to deploy a fake app or read all of the app secrets.  But I don't think having a ranking of secret importance in this doc would be very useful.

File: source/support/rotate-secrets/index.html.md.erb [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+#### `paas-username` and `paas-password`
+
+Change the password by [triggering a password reset][], which gets sent to the govuk-accounts-developers Google Group.
+
+[triggering a password reset]: https://admin.london.cloud.service.gov.uk/password/request-reset
+
+#### `sentry-dsn`
+
+Generate a new DSN from the Sentry project settings page.
```
Comment (jonathanglassman) [inline source/support/rotate-secrets/index.html.md.erb]: Where is the Sentry project settings page?

File: source/support/rotate-secrets/index.html.md.erb [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+- how to rotate secrets if needed
+
+If a secret is compromised, this is a security incident. Follow [the
+incident management process][] as well as rotating the secret.
+
+[the incident management process]: /incident
+
+## Prerequisites
+
+You must [install the gds-cli tool][] to manage secrets.
```
Comment (jonathanglassman) [inline source/support/rotate-secrets/index.html.md.erb]: Are there any other role or permission requirements? For example, would I be right in thinking you must also not be on probation?

File: source/support/rotate-secrets/index.html.md.erb [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+[triggering a password reset]: https://admin.london.cloud.service.gov.uk/password/request-reset
+
+#### `sentry-dsn`
+
+Generate a new DSN from the Sentry project settings page.
+
+#### `slack_webhook_url`
+
+Tell IT Support that our Slack webhook has been leaked, and that we
```
Comment (jonathanglassman) [inline source/support/rotate-secrets/index.html.md.erb]: How? Email? Slack?

File: source/support/rotate-secrets/index.html.md.erb [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+#### `notify-api-key-production` and `notify-api-key-staging`
+
+Generate a new API key from [GOV.UK Notify][], and deactivate the old
+API key.
+
+[GOV.UK Notify]: https://www.notifications.service.gov.uk/
+
+#### `oidc-pepper-production` and `oidc-pepper-staging`
+
+Generate an arbitrary string using `rake secret`.
```
Comment (jonathanglassman) [inline source/support/rotate-secrets/index.html.md.erb:132]: run this in gds-cli? or in some other way?

File: source/support/rotate-secrets/index.html.md.erb [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+#### `secret-key-base-production` and `secret-key-base-staging`
+
+Generate an arbitrary string with `rake secret`.
+
+This secret is used to sign session cookies, and so rotating it will
+log out all users.
+
+#### `zendesk-api-username` and `zendesk-api-key`
+
+Tell User Support that our Zendesk credentials have been leaked, and
```
Comment (jonathanglassman) [inline source/support/rotate-secrets/index.html.md.erb]: How? Email? Slack?

File: source/support/rotate-secrets/index.html.md.erb [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    ```
+    $ gds cd secrets add cd-govuk-tools <pipeline-name>/<secret-name> "<value>"
+    ```
+
+3. If this is a secret used by an app, deploy the app.
+
+## How to rotate the secrets we have
+
+In general you rotate secrets by generating a new value and
+setting the value with the gds-cli. You must use a different method for the following secrets.
```
Comment (jonathanglassman) [inline source/support/rotate-secrets/index.html.md.erb]: So, is the method above a general method and the secrets below are exceptions to that general method?

File: source/support/rotate-secrets/index.html.md.erb (hunk 4 above)
Comment (barrucadu) [reply source/support/rotate-secrets/index.html.md.erb]: Added some steps

File: source/support/rotate-secrets/index.html.md.erb (hunk 5 above)
Comment (barrucadu) [reply source/support/rotate-secrets/index.html.md.erb]: Added that you need production access

File: source/support/rotate-secrets/index.html.md.erb (hunk 6 above)
Comment (barrucadu) [reply source/support/rotate-secrets/index.html.md.erb]: Clarified

File: source/support/rotate-secrets/index.html.md.erb (hunk 8 above)
Comment (barrucadu) [reply source/support/rotate-secrets/index.html.md.erb]: Clarified

File: source/support/rotate-secrets/index.html.md.erb (hunk 9 above)
Comment (barrucadu) [reply source/support/rotate-secrets/index.html.md.erb]: This section is actually a list of all the secrets and how to generate new values.  I've changed the title and removed this paragraph.

File: source/support/rotate-secrets/index.html.md.erb (hunk 7 above)
Comment (barrucadu) [reply source/support/rotate-secrets/index.html.md.erb:132]: `rake` is a tool to run maintenance tasks for the apps, and one of the built-in tasks is this `secret` one which generates a random long string.  So to run this a developer would need a copy of the app code on their machine.

File: source/support/rotate-secrets/index.html.md.erb [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+Generate an arbitrary string using `rake secret`.
+
+This secret is used to [hash account passwords][], and so rotating it
+will invalidate all of the current passwords. This means that users
+will have to reset their password to log in.
+
+If this secret needs to be rotated on short notice, we may want to
+send an email to all users explaining the problem. If it can be
+rotated over a longer period, we could rehash passwords when users log
+in, and only disable the old pepper when most users have been changed.
```
Comment (jonathanglassman) [inline source/support/rotate-secrets/index.html.md.erb]: @barrucadu Should "the old pepper" say something else?

File: source/support/rotate-secrets/index.html.md.erb [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+To list all secrets for the GOV.UK Concourse team, run the following in the gds-cli:
+
+```
+$ gds cd secrets ls cd-govuk-tools
+```
+
+[the daily statistics task]: https://github.com/alphagov/govuk-account-manager-prototype/blob/7e69b8cb8daec1e6e23d32374df3f8a1fa45b477/concourse/pipeline.yml#L249-L271
+[GOV.UK Notify API key]: https://github.com/alphagov/govuk-account-manager-prototype/blob/5443585c3d2dd5d399efc18f126f0bdf831036b5/concourse/tasks/deploy-to-govuk-paas.yml#L59
+
+### Rotating a secret
```
Comment (jonathanglassman) [inline source/support/rotate-secrets/index.html.md.erb]: @barrucadu would make this an H2 so it shows up in the left hand nav

**Review by jonathanglassman — APPROVED** [review]
Approved, with two last comments

File: source/support/rotate-secrets/index.html.md.erb (hunk 10 above)
Comment (barrucadu) [reply source/support/rotate-secrets/index.html.md.erb]: Changed it to "the old value"

**Pull request merged**